agent/xds/clusters.go

```diff
@@ -151,6 +175,85 @@ func (s *Server) makeUpstreamCluster(upstream structs.Upstream, cfgSnap *proxycf
 	return c, nil
 }
 
+func (s *Server) makeUpstreamClustersForDiscoveryChain(
+	upstreamID string,
+	chain *structs.CompiledDiscoveryChain,
+	cfgSnap *proxycfg.ConfigSnapshot,
+) ([]*envoy.Cluster, error) {
+	if chain == nil {
+		panic("chain must be provided")
+	}
+
+	// TODO(rb): make escape hatches work with chains
+
+	var out []*envoy.Cluster
+	for target, node := range chain.GroupResolverNodes {
+		groupResolver := node.GroupResolver
+		// TODO: failover
+		// Failover *DiscoveryFailover `json:",omitempty"` // sad path
+
+		clusterName := makeClusterName(upstreamID, target, cfgSnap.Datacenter)
+		c := &envoy.Cluster{
+			Name:                 clusterName,
+			AltStatName:          clusterName, // TODO(rb): change this?
+			ConnectTimeout:       groupResolver.ConnectTimeout,
+			ClusterDiscoveryType: &envoy.Cluster_Type{Type: envoy.Cluster_EDS},
+			CommonLbConfig: &envoy.Cluster_CommonLbConfig{
+				HealthyPanicThreshold: &envoytype.Percent{
+					Value: 0, // disable panic threshold
+				},
+			},
+			// TODO: adjust load assignment
+			EdsClusterConfig: &envoy.Cluster_EdsClusterConfig{
+				EdsConfig: &envoycore.ConfigSource{
+					ConfigSourceSpecifier: &envoycore.ConfigSource_Ads{
+						Ads: &envoycore.AggregatedConfigSource{},
+					},
+				},
+			},
+			// Having an empty config enables outlier detection with default config.
+			OutlierDetection: &envoycluster.OutlierDetection{},
+		}
+		if groupResolver.Protocol == "http2" || groupResolver.Protocol == "grpc" {
+			c.Http2ProtocolOptions = &envoycore.Http2ProtocolOptions{}
+		}
+
+		// Enable TLS upstream with the configured client certificate.
+		c.TlsContext = &envoyauth.UpstreamTlsContext{
+			CommonTlsContext: makeCommonTLSContext(cfgSnap),
+		}
+
+		out = append(out, c)
+	}
+
+	return out, nil
+}
+
+// makeClusterName returns a string representation that uniquely identifies the
+// cluster in a canonical but human readable way.
+func makeClusterName(upstreamID string, target structs.DiscoveryTarget, currentDatacenter string) string {
```

Comment:
I am wondering if we should unify our cluster naming to be SNI names since thats what they will need to be for gateways. Maybe there is a reason not to 🤷‍♂

Reply:
I'm up for it. It does affect metrics collection or dashboards people may have.

Reply:
We should do it but make a TODO item on our GA list to review metrics and add labelling rules to make the output useful (and ideally backwards compatible) again.